### PR TITLE
have gballoc hl passthrough

### PR DESCRIPTION
**Pull request by anporumb**

### CODE REVIEW

File: common/devdoc/gballoc_hl_passthrough.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
+# gballoc_hl_passthrough requirements
+================
+
+## Overview
+
+gballoc_hl_passthrough is a module that delegates all call of its APIs to the ones from gballoc_ll.
```
Comment (mattdurak) [inline common/devdoc/gballoc_hl_passthrough.md:6]: >gballoc_hl_passthrough [](start = 0, length = 22)

So is this like an interface and there will be other implementations that do something besides passthrough? Sorry, not in the loop on all of this and trying to understand what the purpose of this layer is...

File: common/tests/CMakeLists.txt [hunk 2]
```diff
@@ -29,5 +31,14 @@ if(${run_int_tests})
         build_test_folder(gballoc_ll_mimalloc_int)
         build_test_folder(gballoc_ll_win32heap_int) 
         build_test_folder(gballoc_ll_passthrough_int)
+        
+        #this builds and runs the same test for gballoc_hl but with all the implementations of the underlying gballoc_ll.
+        #this scripts sets a variable (GBALOC_LL_IMPL) that is picked as underlying implementatio of gballoc_ll and consequentially as file name (gballoc_ll_${GBALOC_LL_IMPL})
+        foreach(GBALOC_LL_IMPL ${GBALLOC_LL_TYPE_VALUES})
+        #set(GBALOC_LL_IMPL "PASSTHROUGH")
```
Comment (mattdurak) [inline common/tests/CMakeLists.txt:38]: >#set(GBALOC_LL_IMPL "PASSTHROUGH") [](start = 8, length = 34)

can be removed?

File: common/tests/CMakeLists.txt [hunk 3]
```diff
@@ -29,5 +31,14 @@ if(${run_int_tests})
         build_test_folder(gballoc_ll_mimalloc_int)
         build_test_folder(gballoc_ll_win32heap_int) 
         build_test_folder(gballoc_ll_passthrough_int)
+        
+        #this builds and runs the same test for gballoc_hl but with all the implementations of the underlying gballoc_ll.
+        #this scripts sets a variable (GBALOC_LL_IMPL) that is picked as underlying implementatio of gballoc_ll and consequentially as file name (gballoc_ll_${GBALOC_LL_IMPL})
+        foreach(GBALOC_LL_IMPL ${GBALLOC_LL_TYPE_VALUES})
```
Comment (mattdurak) [inline common/tests/CMakeLists.txt:37]: >GBALLOC_LL_TYPE_VALUES [](start = 33, length = 22)

where is this defined?

File: common/tests/CMakeLists.txt [hunk 4]
```diff
@@ -29,5 +31,14 @@ if(${run_int_tests})
         build_test_folder(gballoc_ll_mimalloc_int)
         build_test_folder(gballoc_ll_win32heap_int) 
         build_test_folder(gballoc_ll_passthrough_int)
+        
+        #this builds and runs the same test for gballoc_hl but with all the implementations of the underlying gballoc_ll.
+        #this scripts sets a variable (GBALOC_LL_IMPL) that is picked as underlying implementatio of gballoc_ll and consequentially as file name (gballoc_ll_${GBALOC_LL_IMPL})
+        foreach(GBALOC_LL_IMPL ${GBALLOC_LL_TYPE_VALUES})
+        #set(GBALOC_LL_IMPL "PASSTHROUGH")
+            STRING(TOLOWER ${GBALOC_LL_IMPL} gballoc_ll_impl_lower)
+            build_test_folder(gballoc_hl_passthrough_int BINARY_DIR _${gballoc_ll_impl_lower})
```
Comment (mattdurak) [inline common/tests/CMakeLists.txt:40]: >_ [](start = 68, length = 1)

leading underscore in directory name intentional?

File: common/devdoc/gballoc_hl_passthrough.md (hunk 1 above)
Comment (anporumb) [reply common/devdoc/gballoc_hl_passthrough.md:6]: We need/want an abstraction that "does nothing" for those times when we don't need to compute anything

And there's going to be another implementation that tracks metrics. 

A user is only supposed to be using the HL layer (which will become the gballoc we know).

---
In reply to: [458464922](https://github.com/Azure/azure-c-pal/pull/12#discussion_r458464922) [](ancestors = 458464922)

File: common/tests/CMakeLists.txt (hunk 3 above)
Comment (anporumb) [reply common/tests/CMakeLists.txt:37]: This is defined in the master CMakeLists.txt of this repo. It contains all the values that the LL can have (and it also checks that the user didn't pass some unknown value). Right now this list has 3 values: "passthrough (directly to malloc) win32heap and mimalloc".

The purpose here being to check that every implementation of the HL (at the moment 1) can work with any implementation of the LL layer - composability.

---
In reply to: [458466154](https://github.com/Azure/azure-c-pal/pull/12#discussion_r458466154) [](ancestors = 458466154)

File: common/tests/CMakeLists.txt (hunk 2 above)
Comment (anporumb) [reply common/tests/CMakeLists.txt:38]: Yes, leftovers from CMake debug. Thanks for noticing. :)

---
In reply to: [458465947](https://github.com/Azure/azure-c-pal/pull/12#discussion_r458465947) [](ancestors = 458465947)

Comment (anporumb) [reply common/tests/CMakeLists.txt:38]: Removed.

---
In reply to: [458501198](https://github.com/Azure/azure-c-pal/pull/12#discussion_r458501198) [](ancestors = 458501198,458465947)

File: common/tests/CMakeLists.txt (hunk 4 above)
Comment (anporumb) [reply common/tests/CMakeLists.txt:40]: yeah, I was not aware that it shouldn't be done. Results in a folder such as this: "common\tests\gballoc_hl_passthrough_int\_mimalloc\dll".


---
In reply to: [458466489](https://github.com/Azure/azure-c-pal/pull/12#discussion_r458466489) [](ancestors = 458466489)

Comment (anporumb) [reply common/tests/CMakeLists.txt:40]: Maybe I can do better - moment. :)

---
In reply to: [458501872](https://github.com/Azure/azure-c-pal/pull/12#discussion_r458501872) [](ancestors = 458501872,458466489)

Comment (anporumb) [reply common/tests/CMakeLists.txt:40]: alright, changed CMakeLists.txt to add an "_" be convention in some other place.  NOw folders on the drive look like: "tests\gballoc_hl_passthrough_int\mimalloc\dll"

---
In reply to: [458501872](https://github.com/Azure/azure-c-pal/pull/12#discussion_r458501872) [](ancestors = 458501872,458466489)

**Pull request merged**